File: Tutorials/Codes/Modern_Control_Course/Chapter18/Lesson5/Chapter18_Lesson5.m
% Chapter18_Lesson5.m
% Interpretation of Jordan form in control applications.
%
% Required for the main analysis:
%   MATLAB base functions: expm, rank
% Recommended:
%   Control System Toolbox: ctrb, obsv, ss, initial
%   Symbolic Math Toolbox: jordan, sym
% Optional:
%   Simulink for automatic construction of a State-Space block model

clear; clc; close all;

lambda = -1;
A = [lambda 1 0;
     0 lambda 1;
     0 0 lambda];

fprintf('A = Jordan block J_3(-1):\n');
disp(A);

t = 2;
E_closed = exp(lambda*t) * [1 t t^2/factorial(2);
                            0 1 t;
                            0 0 1];
E_numeric = expm(A*t);

fprintf('Closed-form exp(A t) at t=2:\n');
disp(E_closed);
fprintf('MATLAB expm(A t) at t=2:\n');
disp(E_numeric);
fprintf('Max absolute difference = %.3e\n', max(abs(E_closed - E_numeric), [], 'all'));

B_good = [0; 0; 1];
B_bad = [1; 0; 0];

Wc_good = ctrb(A, B_good);
Wc_bad = ctrb(A, B_bad);

fprintf('\nControllability rank with B_good = %d of %d\n', rank(Wc_good), size(A,1));
disp(Wc_good);
fprintf('Controllability rank with B_bad = %d of %d\n', rank(Wc_bad), size(A,1));
disp(Wc_bad);

C_good = [1 0 0];
C_bad = [0 0 1];

Wo_good = obsv(A, C_good);
Wo_bad = obsv(A, C_bad);

fprintf('\nObservability rank with C_good = %d of %d\n', rank(Wo_good), size(A,1));
disp(Wo_good);
fprintf('Observability rank with C_bad = %d of %d\n', rank(Wo_bad), size(A,1));
disp(Wo_bad);

% Free-response comparison for a generalized eigenvector initial condition.
x0 = [0; 0; 1];
time = linspace(0, 6, 121);
X = zeros(numel(time), 3);
for k = 1:numel(time)
    X(k,:) = (expm(A*time(k))*x0).';
end

figure('Name','Chapter18 Lesson5 Jordan-chain response');
plot(time, X, 'LineWidth', 1.5);
grid on;
xlabel('time');
ylabel('state components');
legend('x_1','x_2','x_3','Location','best');
title('Polynomial-times-exponential response of a Jordan chain');

% Symbolic Jordan decomposition when Symbolic Math Toolbox is available.
if license('test','Symbolic_Toolbox')
    fprintf('\nSymbolic Jordan form of A:\n');
    Js = jordan(sym(A));
    disp(Js);
end

% Minimal Simulink construction: continuous State-Space block with A,B,C,D.
if license('test','Simulink')
    mdl = 'Chapter18_Lesson5_Simulink';
    if bdIsLoaded(mdl)
        close_system(mdl, 0);
    end
    new_system(mdl);
    open_system(mdl);
    add_block('simulink/Continuous/State-Space', [mdl '/Jordan_Chain_State_Space']);
    set_param([mdl '/Jordan_Chain_State_Space'], ...
        'A', mat2str(A), ...
        'B', mat2str(B_good), ...
        'C', mat2str(C_good), ...
        'D', '0', ...
        'X0', mat2str(x0));
    add_block('simulink/Sinks/Scope', [mdl '/Scope']);
    add_line(mdl, 'Jordan_Chain_State_Space/1', 'Scope/1');
    save_system(mdl);
    fprintf('Created Simulink model: %s.slx\n', mdl);
end

File: Tutorials/Codes/Modern_Control_Course/Chapter18/Lesson5/Chapter18_Lesson5.py
"""
Chapter18_Lesson5.py
Interpretation of Jordan form in control applications.

Required packages:
    numpy
    scipy

This script demonstrates:
1. matrix exponentials of Jordan blocks,
2. polynomial-times-exponential response terms,
3. controllability/observability rank tests in Jordan coordinates,
4. how input/output placement along a Jordan chain affects control authority and sensing.
"""

from __future__ import annotations

import math
from typing import Iterable

import numpy as np
from scipy.linalg import expm


def jordan_block(lam: float, size: int) -> np.ndarray:
    """Return one Jordan block J = lam*I + N with ones on the superdiagonal."""
    if size <= 0:
        raise ValueError("size must be positive")
    J = lam * np.eye(size)
    for i in range(size - 1):
        J[i, i + 1] = 1.0
    return J


def jordan_block_exponential(lam: float, size: int, t: float) -> np.ndarray:
    """Closed-form exp(J t) for one Jordan block."""
    E = np.zeros((size, size))
    scale = math.exp(lam * t)
    for i in range(size):
        for j in range(i, size):
            power = j - i
            E[i, j] = scale * (t ** power) / math.factorial(power)
    return E


def controllability_matrix(A: np.ndarray, B: np.ndarray) -> np.ndarray:
    """Construct Ctrb = [B, AB, ..., A^(n-1)B]."""
    n = A.shape[0]
    blocks = []
    Ak = np.eye(n)
    for _ in range(n):
        blocks.append(Ak @ B)
        Ak = A @ Ak
    return np.hstack(blocks)


def observability_matrix(A: np.ndarray, C: np.ndarray) -> np.ndarray:
    """Construct Obsv = [C; CA; ...; CA^(n-1)]."""
    n = A.shape[0]
    blocks = []
    Ak = np.eye(n)
    for _ in range(n):
        blocks.append(C @ Ak)
        Ak = Ak @ A
    return np.vstack(blocks)


def numerical_rank(M: np.ndarray, tol: float = 1e-10) -> int:
    """Numerical rank using singular values."""
    return int(np.linalg.matrix_rank(M, tol=tol))


def print_modal_response(A: np.ndarray, x0: Iterable[float], times: Iterable[float]) -> None:
    x0_vec = np.asarray(list(x0), dtype=float).reshape((-1, 1))
    print("\nFree response x(t) = exp(A t) x0")
    for t in times:
        xt = expm(A * t) @ x0_vec
        print(f"t={t:4.1f}: x(t)^T = {xt.ravel()}")


def main() -> None:
    # One defective eigenvalue: a length-3 Jordan chain at lambda = -1.
    lam = -1.0
    A = jordan_block(lam, 3)
    print("A = Jordan block J_3(-1):")
    print(A)

    t = 2.0
    closed = jordan_block_exponential(lam, 3, t)
    numeric = expm(A * t)
    print("\nClosed-form exp(A t) at t=2:")
    print(closed)
    print("\nSciPy expm(A t) at t=2:")
    print(numeric)
    print("\nMax absolute difference:", np.max(np.abs(closed - numeric)))

    # Control authority depends on where the input enters the chain.
    B_good = np.array([[0.0], [0.0], [1.0]])  # acts on tail of chain, reaches all generalized directions
    B_bad = np.array([[1.0], [0.0], [0.0]])   # acts only on head direction

    for name, B in [("B_good", B_good), ("B_bad", B_bad)]:
        Wc = controllability_matrix(A, B)
        print(f"\n{name} =\n{B}")
        print("Controllability matrix =")
        print(Wc)
        print("rank =", numerical_rank(Wc), "of", A.shape[0])

    # Sensing also depends on output placement along the chain.
    C_good = np.array([[1.0, 0.0, 0.0]])  # measures head; chain derivatives reveal all components
    C_bad = np.array([[0.0, 0.0, 1.0]])   # measures tail only

    for name, C in [("C_good", C_good), ("C_bad", C_bad)]:
        Wo = observability_matrix(A, C)
        print(f"\n{name} = {C}")
        print("Observability matrix =")
        print(Wo)
        print("rank =", numerical_rank(Wo), "of", A.shape[0])

    print_modal_response(A, x0=[0.0, 0.0, 1.0], times=[0, 0.5, 1, 2, 4, 6])


if __name__ == "__main__":
    main()
